packages/@aws-cdk/aws-appmesh/lib/virtual-router.ts

```diff
@@ -0,0 +1,277 @@
+   */
+  public addRoutes(ids: string[], props: RouteBaseProps[]): Route[] {
+    const routes = new Array<Route>();
+
+    if (ids.length < 1 || props.length < 1) {
+      throw new Error('When adding routes, IDs and Route Properties cannot be empty.');
+    }
+    if (ids.length !== props.length) {
+      throw new Error('Routes must have the same number of IDs and RouteProps.');
+    }
+
+    for (let i = 0; i < ids.length; i++) {
+      const route = new Route(this, ids[i], {
+        routeName: ids[i],
+        mesh: this.mesh,
+        virtualRouter: this,
+        routeTargets: props[i].routeTargets,
+        isHttpRoute: props[i].isHttpRoute,
+        prefix: props[i].prefix,
+      });
+      routes.push(route);
+    }
+
+    return routes;
+  }
+}
+
+/**
+ * The properties used when creating a new VritualRouter
+ */
+export interface VirtualRouterProps extends VirtualRouterBaseProps {
+  /**
+   * The AppMesh mesh the VirtualRouter belongs to
+   */
+  readonly mesh: IMesh;
+}
+
+export class VirtualRouter extends VirtualRouterBase {
+  /**
+   * Import an existing VirtualRouter given an ARN
+   */
+  public static fromVirtualRouterArn(scope: cdk.Construct, id: string, virtualRouterArn: string): IVirtualRouter {
+    return new ImportedVirtualRouter(scope, id, { virtualRouterArn });
+  }
+
+  /**
+   * Import an existing VirtualRouter given names
+   */
+  public static fromVirtualRouterName(scope: cdk.Construct, id: string, meshName: string, virtualRouterName: string): IVirtualRouter {
+    return new ImportedVirtualRouter(scope, id, { meshName, virtualRouterName });
+  }
+
+  /**
+   * The name of the VirtualRouter
+   */
+  public readonly virtualRouterName: string;
+
+  /**
+   * The Amazon Resource Name (ARN) for the VirtualRouter
+   */
+  public readonly virtualRouterArn: string;
+
+  /**
+   * The AppMesh mesh the VirtualRouter belongs to
+   */
+  public readonly mesh: IMesh;
+
+  private readonly listeners = new Array<CfnVirtualRouter.VirtualRouterListenerProperty>();
+
+  constructor(scope: cdk.Construct, id: string, props: VirtualRouterProps) {
+    super(scope, id, {
+      physicalName: props.virtualRouterName || cdk.Lazy.stringValue({ produce: () => this.node.uniqueId })
+    });
+
+    this.mesh = props.mesh;
+
+    this.addListeners(props);
+
+    const router = new CfnVirtualRouter(this, 'Resource', {
+      virtualRouterName: this.physicalName,
+      meshName: this.mesh.meshName,
+      spec: {
+        listeners: this.listeners,
+      },
+    });
+
+    this.virtualRouterName = this.getResourceNameAttribute(router.attrVirtualRouterName);
+    this.virtualRouterArn = this.getResourceArnAttribute(router.ref, {
+      service: 'appmesh',
+      resource: `mesh/${props.mesh.meshName}/virtualRouter`,
+      resourceName: this.physicalName,
+    });
+  }
+
+  /**
+   * Add listeners to the router, such as ports and protocols
+   */
+  private addListeners(props: VirtualRouterBaseProps) {
```

Comment:
Could the argument be `PortMappingProps` instead, since it appears the `virtualRouterName` in the `VirutalRouterBaseProps` is never used here?